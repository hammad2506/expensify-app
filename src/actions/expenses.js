import database from '../firebase/firebase';

export const addExpense = (expense) => (
    {
    type: 'ADD_EXPENSE',
    expense
    }
);

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.user;
        const {description = '', amount = 0, note = '', createdAt = undefined} = expenseData;
        const expense = {description, amount, note, createdAt};
        return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

export const setExpenses = (expenses) => (
    {
        type: 'SET_EXPENSES',
        expenses
    }
);

export const startSetExpenses = () => { 
    return (dispatch, getState) => {
        const uid = getState().auth.user;
        return database.ref(`users/${uid}/expenses`).once('value').then((dataSnapshot) => {
            const expenses = [];
            dataSnapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            }); 
            dispatch(setExpenses(expenses));
        })
    }
}


export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ({id}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.user;
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>{
            dispatch(removeExpense({id}));
        });
    }
}

export const editExpense = ( id, updates= {} ) => ({
    type: 'EDIT_EXPENSE',
    id, 
    updates
});

export const startEditExpense = ( id, updates= {} ) => {
    return (dispatch, getState) => {
        const uid = getState().auth.user;
        return database.ref(`users/${uid}/expenses/${id}`).update({
            ...updates
        }).then(()=>{
            dispatch(editExpense(id, updates));
        });
    }
}
