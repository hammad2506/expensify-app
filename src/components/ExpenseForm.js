import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends React.Component {

    constructor(props){
 
        super(props);
        this.state= {
            description: '',
            amount: '',
            note: '',
            createdAt: moment(),
            focused: false,
            error: ''
        }
    }

    componentDidMount(){
        if(this.props.expense){
            this.setState(()=>({
                description: this.props.expense.description,
                amount: (this.props.expense.amount/100).toFixed(2),
                note: this.props.expense.note,
                createdAt: moment(this.props.expense.createdAt)
            }));
        }
    }    

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
        
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}));
        }
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState(() => ({ createdAt }));
        }
        
    }

    onFormSubmit = (e) => {
        e.preventDefault();

        if(!this.state.amount || !this.state.description){
            this.setState(() => ({error: "Please enter the description and amount"}));
        } else{
            this.setState(() => ({error: ""}));
            this.props.onSubmit({
                description: this.state.description, 
                note: this.state.note, 
                amount: parseFloat(this.state.amount)*100, 
                createdAt: this.state.createdAt.valueOf()
            })
        }
    }

    render(){
        return(
                <form className='form' onSubmit={this.onFormSubmit}>
                    {this.state.error && <p>{this.state.error}</p>}
                    <input
                        className='text-input' 
                        type='text' placeholder='Description' 
                        value={this.state.description} onChange={this.onDescriptionChange}
                    />
                    <input 
                        className='text-input'
                        type='text' placeholder='Amount' 
                        value={this.state.amount} onChange={this.onAmountChange}
                    />
                    <div>
                        <SingleDatePicker 
                            date={this.state.createdAt} 
                            numberOfMonths={1} 
                            onDateChange= {this.onDateChange}
                            focused={this.state.focused} 
                            isOutsideRange={() => false} 
                            onFocusChange = {({ focused }) => (this.setState( {focused} ))}
                            showDefaultInputIcon={true}
                        />
                    </div>  
                    <textarea
                        className='textarea' 
                        placeholder='Notes' 
                        value={this.state.note} onChange={this.onNoteChange}>
                    </textarea>
                    <div>
                        <button className='button'>
                            {!this.props.expense ? `Add Expense` : `Save Expense`}
                        </button>
                    </div>
                </form>
        );
    }
}

export default ExpenseForm;