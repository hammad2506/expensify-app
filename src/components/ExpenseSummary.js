import React from 'react';
import numeral from 'numeral';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import visibleExpenses from '../selectors/getVisibleExpenses';
import getTotalAmount from '../selectors/get-total';

export const ExpenseSummary = ({ expenseCount, expenseAmount }) => {
    const expenseWord = expenseCount>1 ? 'expenses' : 'expense';
    const amount = numeral(expenseAmount/100).format('$0,0.00'); 
    return (
        <div className='page-header'>
            <div className='content-container'>
                {   expenseCount>0 ? 
                    <h1 className='page-header__title'>Viewing <span>{expenseCount}</span> {expenseWord} totalling <span>{amount}</span></h1> : 
                    undefined 
                }
                <Link className='button page-header__actions' to='/add'>Add Expense</Link>    
            </div>
        </div>      
    )
}

const mapStateToProps = (state) => {
    const filteredExpenses = visibleExpenses(state.expenses, state.filters)
    return {
        expenseCount: filteredExpenses.length,
        expenseAmount: getTotalAmount(filteredExpenses)
    }
}

export default connect(mapStateToProps)(ExpenseSummary);