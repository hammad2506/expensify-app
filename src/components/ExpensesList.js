import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import filterExpenses from '../selectors/getVisibleExpenses'

export const ExpensesList = (props) => {
    return (
        <div>
            {props.expenses.map((expense) => (
                <ExpenseListItem key={expense.id} {...expense} />
            ))}
        </div>
    );
}

const mapStateToProps = (state) => ({
    expenses: filterExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpensesList);