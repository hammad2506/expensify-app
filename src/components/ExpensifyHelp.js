import React from 'react';

const ExpensifyHelp = () => (
    <div>
        <h3>Welcome to the help page</h3>
    </div>
);

export default ExpensifyHelp;