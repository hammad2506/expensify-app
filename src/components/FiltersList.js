import React from 'react';
import { connect } from 'react-redux';
import { changeText, changeSortBy, changeStartDate, changeEndDate } from '../actions/filters';
import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';


export class FiltersList extends React.Component{

    constructor(props){
        super(props);

        this.state = {
            focusedInput: null
        }
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.changeStartDate(startDate);
        this.props.changeEndDate(endDate);
    };

    onTextChange = (e) => {
        const text = e.target.value;
        this.props.changeText(text);
    }

    onSortByChange = (e) => {
        const value = e.target.value;
        if(value === 'date' || value === 'amount'){
            this.props.changeSortBy(value);
        }
    }


    render() {
        return(
            <div className='content-container'>
                <div className='input-group'>
                    <div className='input-group__item'>
                        <input className='text-input' type="text" value={this.props.filters.text} 
                        onChange={this.onTextChange} placeholder='Enter search text here'/>
                    </div>
                    <div className='input-group__item'>
                        <select className='select' value={this.props.filters.sortBy} 
                        onChange={this.onSortByChange}>
                        <option value='date'>Date</option>
                        <option value='amount'>Amount</option>
                        </select>
                    </div>
                    <div>
                        <DateRangePicker 
                                startDate={this.props.filters.startDate}
                                endDate={this.props.filters.endDate}
                                isOutsideRange={() => false}
                                onDatesChange={this.onDatesChange}
                                focusedInput={this.state.focusedInput}
                                onFocusChange={focusedInput => this.setState({ focusedInput })}
                                numberOfMonths={1}
                                showClearDates={true}
                            />
                    </div>
                </div>
             </div>
        )
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    changeText : (text) => dispatch(changeText(text)),
    changeSortBy : (value) => dispatch(changeSortBy(value)),
    changeStartDate : (startDate) => dispatch(changeStartDate(startDate)),
    changeEndDate : (endDate) => dispatch(changeEndDate(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(FiltersList);