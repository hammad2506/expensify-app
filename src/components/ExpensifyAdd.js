import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export class ExpensifyAdd extends React.Component {

    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push('/dashboard');
    }

    render(){
        return (
            <div>
                <h3>Welcome to the Add page</h3>
                <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        );
    }  
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpense : (expense) => dispatch(startAddExpense(expense)) 
});

export default connect(undefined, mapDispatchToProps)(ExpensifyAdd);