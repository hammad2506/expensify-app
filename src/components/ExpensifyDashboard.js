import React from 'react';
import ExpensesList from './ExpensesList';
import ExpenseSummary from './ExpenseSummary';
import FiltersList from './FiltersList';
import { connect } from 'react-redux';

const ExpensifyDashboard = (props) => {
    return(
    <div>
        <ExpenseSummary />
        <FiltersList />
        <ExpensesList />
    </div>
    );
};

const mapStateToProps = (state) => (
    {
        expenses: state.expenses
    }
);


export default connect(mapStateToProps)(ExpensifyDashboard);