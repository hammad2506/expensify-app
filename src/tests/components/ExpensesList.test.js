import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesList } from '../../components/ExpensesList';
import expenses from '../dummyData';

test("Testing Expense List with expenses", () => {
    const wrapper =  shallow(<ExpensesList expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
});

test("Testing Expense List with NO expenses", () => {
    const wrapper =  shallow(<ExpensesList expenses={[]}/>);
    expect(wrapper).toMatchSnapshot();
});