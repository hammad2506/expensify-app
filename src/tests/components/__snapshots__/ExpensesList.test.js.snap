// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Expense List with NO expenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No Expenses
      </span>
    </div>
  </div>
</div>
`;

exports[`Testing Expense List with expenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <ExpenseListItem
      amount={2000}
      createdAt={0}
      description="Bought Gas"
      id="1"
      key="1"
      note="Getting expensive"
    />
    <ExpenseListItem
      amount={3500}
      createdAt={259200000}
      description="Electricity"
      id="2"
      key="2"
      note="October month"
    />
    <ExpenseListItem
      amount={100}
      createdAt={432000000}
      description="Phone Bill"
      id="3"
      key="3"
      note="YOLO"
    />
  </div>
</div>
`;
